File: backend/src/HIS.Core/Entities/SpecimenImage.cs
namespace HIS.Core.Entities;

/// <summary>
/// Ảnh bệnh phẩm gắn vào kết quả XN / giải phẫu bệnh / yêu cầu dịch vụ.
/// Nguồn: chụp kính hiển vi (microscope), upload thủ công từ web, hoặc webcam (getUserMedia).
/// Tích hợp camera device-side native (SDK máy kính) gắn vào đây sau khi có phần cứng.
/// Issues: #134 (gắn ảnh KQ), #133/#113 (web-upload + webcam fallback).
/// </summary>
public class SpecimenImage : BaseEntity
{
    /// <summary>
    /// ID phiếu yêu cầu GPB (PathologyRequest.Id).
    /// Dùng khi gắn ảnh trực tiếp vào phiếu GPB từ drawer (không cần có kết quả trước).
    /// </summary>
    public Guid? PathologyRequestId { get; set; }

    /// <summary>
    /// ID kết quả giải phẫu bệnh (PathologyResult.Id).
    /// Null nếu gắn vào PathologyRequest hoặc ServiceRequestDetail.
    /// </summary>
    public Guid? PathologyResultId { get; set; }

    /// <summary>
    /// ID chi tiết yêu cầu dịch vụ (ServiceRequestDetail.Id) — dùng cho KQ XN.
    /// Null nếu gắn vào GPB.
    /// </summary>
    public Guid? ServiceRequestDetailId { get; set; }

    /// <summary>
    /// ID yêu cầu dịch vụ cha (ServiceRequest.Id) — cho phép query nhanh theo order.
    /// Denormalized để tránh JOIN khi list ảnh theo đơn.
    /// </summary>
    public Guid? ServiceRequestId { get; set; }

    /// <summary>
    /// Đường dẫn file (tương đối, vd: /api/specimen-image/file/{id}) hoặc URL tuyệt đối (R2).
    /// </summary>
    public string? ImagePath { get; set; }

    /// <summary>
    /// Thumbnail base64 nhỏ (≤200×200 px) để render nhanh trong worklist — KHÔNG lưu ảnh full.
    /// </summary>
    public string? ThumbnailBase64 { get; set; }

    /// <summary>
    /// Chú thích cho ảnh (vd: "Tế bào ung thư biểu mô — x400").
    /// </summary>
    public string? Caption { get; set; }

    /// <summary>
    /// Thời điểm chụp/upload.
    /// </summary>
    public DateTime CapturedAt { get; set; }

    /// <summary>
    /// Nguồn ảnh: microscope | manual | webcam.
    /// - microscope: ảnh từ camera gắn kính hiển vi (device-side, cần SDK — defer đến khi có phần cứng).
    /// - manual: upload file thủ công từ web (đã làm trong #134/#133).
    /// - webcam: chụp từ webcam qua getUserMedia (đã làm trong #133).
    /// </summary>
    public string Source { get; set; } = "manual";

    /// <summary>Tên file gốc khi upload.</summary>
    public string? FileName { get; set; }

    /// <summary>MIME type (image/jpeg, image/png, image/webp...).</summary>
    public string? MimeType { get; set; }

    /// <summary>Kích thước file tính bằng byte.</summary>
    public long FileSize { get; set; }

    /// <summary>Độ phóng đại kính hiển vi (x40, x100, x400...) — điền khi source=microscope.</summary>
    public string? Magnification { get; set; }

    /// <summary>Thứ tự hiển thị trong gallery.</summary>
    public int SortOrder { get; set; }

    /// <summary>Đính kèm vào phiếu in kết quả hay không.</summary>
    public bool IncludeInReport { get; set; } = true;
}

File: frontend/src/api/specimenImage.ts
/**
 * API client — SpecimenImage (ảnh kính hiển vi / upload ảnh KQ XN / GPB)
 * Issues: #134 (gắn ảnh KQ), #133/#113 (web-upload + webcam fallback)
 *
 * Điểm tích hợp camera device-side (SDK máy kính) trong tương lai:
 * - POST /api/specimen-image/upload với source="microscope" + magnification
 */
import { apiClient } from '../services/apiClient';

// ─── Types ────────────────────────────────────────────────────────

export interface SpecimenImage {
  id: string;
  imagePath: string;
  fileName?: string;
  mimeType?: string;
  fileSize: number;
  caption?: string;
  source: 'microscope' | 'manual' | 'webcam';
  magnification?: string;
  sortOrder: number;
  includeInReport: boolean;
  capturedAt: string;
}

export interface UploadImageResult {
  id: string;
  imagePath: string;
  fileName?: string;
  fileSize: number;
  caption?: string;
  source: string;
  capturedAt: string;
}

// ─── Upload file (manual / webcam file) ──────────────────────────

export interface UploadFileParams {
  file: File;
  pathologyResultId?: string;
  serviceRequestDetailId?: string;
  serviceRequestId?: string;
  caption?: string;
  magnification?: string;
  source?: 'manual' | 'webcam' | 'microscope';
}

export const uploadSpecimenImageFile = async (
  params: UploadFileParams
): Promise<UploadImageResult> => {
  const form = new FormData();
  form.append('file', params.file);
  if (params.pathologyResultId) form.append('pathologyResultId', params.pathologyResultId);
  if (params.serviceRequestDetailId) form.append('serviceRequestDetailId', params.serviceRequestDetailId);
  if (params.serviceRequestId) form.append('serviceRequestId', params.serviceRequestId);
  if (params.caption) form.append('caption', params.caption);
  if (params.magnification) form.append('magnification', params.magnification);
  form.append('source', params.source ?? 'manual');

  const res = await apiClient.post<UploadImageResult>('/specimen-image/upload', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return res.data;
};

// ─── Upload base64 (từ webcam getUserMedia) ───────────────────────

export interface UploadBase64Params {
  base64Data: string;       // data:image/jpeg;base64,... hoặc raw base64
  mimeType: string;
  pathologyResultId?: string;
  serviceRequestDetailId?: string;
  serviceRequestId?: string;
  caption?: string;
  magnification?: string;
  source?: 'webcam' | 'microscope';
}

export const uploadSpecimenImageBase64 = async (
  params: UploadBase64Params
): Promise<UploadImageResult> => {
  const res = await apiClient.post<UploadImageResult>('/specimen-image/upload-base64', {
    base64Data: params.base64Data,
    mimeType: params.mimeType,
    pathologyResultId: params.pathologyResultId ?? null,
    serviceRequestDetailId: params.serviceRequestDetailId ?? null,
    serviceRequestId: params.serviceRequestId ?? null,
    caption: params.caption ?? null,
    magnification: params.magnification ?? null,
    source: params.source ?? 'webcam',
  });
  return res.data;
};

// ─── List ─────────────────────────────────────────────────────────

export const listByPathologyResult = async (resultId: string): Promise<SpecimenImage[]> => {
  try {
    const res = await apiClient.get<SpecimenImage[]>(`/specimen-image/by-pathology-result/${resultId}`);
    return res.data ?? [];
  } catch {
    console.warn('[specimenImage] listByPathologyResult failed');
    return [];
  }
};

export const listByServiceDetail = async (detailId: string): Promise<SpecimenImage[]> => {
  try {
    const res = await apiClient.get<SpecimenImage[]>(`/specimen-image/by-service-detail/${detailId}`);
    return res.data ?? [];
  } catch {
    console.warn('[specimenImage] listByServiceDetail failed');
    return [];
  }
};

export const listByServiceRequest = async (requestId: string): Promise<SpecimenImage[]> => {
  try {
    const res = await apiClient.get<SpecimenImage[]>(`/specimen-image/by-service-request/${requestId}`);
    return res.data ?? [];
  } catch {
    console.warn('[specimenImage] listByServiceRequest failed');
    return [];
  }
};

// ─── Update / Delete ──────────────────────────────────────────────

export const updateSpecimenImage = async (
  id: string,
  data: { caption?: string; magnification?: string; includeInReport?: boolean; sortOrder?: number }
): Promise<void> => {
  await apiClient.put(`/specimen-image/${id}`, data);
};

export const deleteSpecimenImage = async (id: string): Promise<void> => {
  await apiClient.delete(`/specimen-image/${id}`);
};
